### PR TITLE
add Tensor Parallelism to Ovis-Image

### PR DESCRIPTION
**Pull request by ZANMANGLOOPYE**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose
Feature for #1217 .
Add Tensor Parallelism to Ovis-Image.
## Test Plan
python text_to_image.py --model AIDC-AI/Ovis-Image-7B --prompt "a cup of coffee on the table" --negative_prompt "ugly, unclear" --cfg_scale 4.0 --num_inference_steps 50 --output "tp_enabled.png" --tensor_parallel_size 4
## Test Result
Image size: 1024x1024
| TP Size | Time | Generated Image |
| :--- | :---: | :--- |
| TP=1 | 12306.82ms |  |
| TP=2 | 6979.08ms |  |
| TP=4 | 4246.40ms |  |
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by hsliuustc0106** [comment]
tp3?

**Comment by ZANMANGLOOPYE** [comment]
> tp3?
Sorry, I made a mistake. It should be corrected as TP = 4.

File: vllm_omni/diffusion/models/ovis_image/ovis_image_transformer.py [hunk 1]
```diff
@@ -520,25 +637,70 @@ def load_weights(self, weights: Iterable[tuple[str, torch.Tensor]]) -> set[str]:
         ]
 
         params_dict = dict(self.named_parameters())
+        loaded_params: set[str] = set()
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/ovis_image/ovis_image_transformer.py:643]: The original code had buffer loading for XIELU's `beta` and `eps` parameters:

```python
# we need to load the buffers for beta and eps (XIELU)
for name, buffer in self.named_buffers():
    if name.endswith(".beta") or name.endswith(".eps"):
        params_dict[name] = buffer
```

This was removed in this PR. The Flux transformer (`flux_transformer.py`) still has this code. Was this intentional? If XIELU is used (or could be used in the future) with Ovis-Image models, this will silently fail to load those buffers.

File: vllm_omni/diffusion/models/ovis_image/ovis_image_transformer.py [hunk 2]
```diff
@@ -151,13 +260,14 @@ def forward(
             value,
         )
         hidden_states = hidden_states.flatten(2, 3)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/ovis_image/ovis_image_transformer.py:263]: The original code had `hidden_states = hidden_states.to(query.dtype)` here. This line was removed — is this intentional? It served as a safety cast (e.g., if the attention backend returns a different dtype). The Flux transformer still has this cast at the same location.

**Review by lishunyang12 — COMMENTED** [review]
Thanks for the contribution, a few comments below.

File: vllm_omni/diffusion/models/ovis_image/ovis_image_transformer.py [hunk 3]
```diff
@@ -396,8 +511,8 @@ def __init__(
         self.timestep_embedder = TimestepEmbedding(in_channels=256, time_embed_dim=self.inner_dim)
 
         self.context_embedder_norm = RMSNorm(joint_attention_dim, eps=1e-6)
-        self.context_embedder = nn.Linear(joint_attention_dim, self.inner_dim)
-        self.x_embedder = nn.Linear(in_channels, self.inner_dim)
+        self.context_embedder = ColumnParallelLinear(joint_attention_dim, self.inner_dim, bias=True, gather_output=True)
+        self.x_embedder = ColumnParallelLinear(in_channels, self.inner_dim, bias=True, gather_output=True)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/ovis_image/ovis_image_transformer.py]: Using `ColumnParallelLinear` with `gather_output=True` for `context_embedder` and `x_embedder` introduces an all-gather communication step that the original `nn.Linear` didn't have. For an equivalent to the original behavior, a plain `nn.Linear` is sufficient here (the weights are small relative to the transformer blocks). The Flux model intentionally keeps these as `nn.Linear` for this reason.

If you do want TP here for memory savings, consider using `ReplicatedLinear` instead, which avoids the all-gather cost.

File: vllm_omni/diffusion/models/ovis_image/ovis_image_transformer.py [hunk 4]
```diff
@@ -520,25 +637,70 @@ def load_weights(self, weights: Iterable[tuple[str, torch.Tensor]]) -> set[str]:
...
+                        param.weight_loader(param, loaded_weight, shard_id)
+                        loaded_params.add(target_param_name)
+                        found_stacked = True
+                        break
+            if found_stacked:
+                continue
+
+            # 4. Standard parameter loading
+            if name not in params_dict:
+                continue
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/ovis_image/ovis_image_transformer.py:704]: Silently skipping parameters that aren't in `params_dict` with `continue` can hide real bugs (e.g., a weight name mapping error). The original code would raise a `KeyError` if a parameter name wasn't found, which is a much safer default. Consider at least logging a warning here:

```python
if name not in params_dict:
    logger.warning("Unexpected parameter in checkpoint: %s", name)
    continue
```

File: vllm_omni/diffusion/models/ovis_image/ovis_image_transformer.py [hunk 5]
```diff
@@ -37,6 +47,103 @@
...
+
+class SwiGLU(nn.Module):
+    def __init__(self, dim_in: int, inner_dim: int, bias: bool = True):
+        super().__init__()
+        self.proj = ColumnParallelLinear(dim_in, inner_dim * 2, bias=bias, gather_output=False)
+        self.activation = nn.SiLU()
+
+    def forward(self, hidden_states):
+        hidden_states, _ = self.proj(hidden_states)
+        hidden_states, gate = hidden_states.chunk(2, dim=-1)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/ovis_image/ovis_image_transformer.py:103]: Nit: `ColumnParallelLinear` with `gather_output=False` outputs `[..., inner_dim * 2 / tp_size]`. Then `.chunk(2, dim=-1)` splits it into two tensors of size `inner_dim / tp_size`. This works correctly because the weight layout is `[hidden, gate]` interleaved within the TP shard.

However, this only works because the custom weight loader (lines 659-682) manually re-interleaves the weights per shard. It might be worth adding a brief comment here explaining the expected layout, e.g.:

```python
# Weight layout: [hidden_local, gate_local] interleaved within each TP shard
hidden_states, gate = hidden_states.chunk(2, dim=-1)
```

File: vllm_omni/diffusion/models/ovis_image/ovis_image_transformer.py [hunk 6]
```diff
@@ -421,7 +536,9 @@ def __init__(
             ]
         )
         self.norm_out = AdaLayerNormContinuous(self.inner_dim, self.inner_dim, elementwise_affine=False, eps=1e-6)
-        self.proj_out = nn.Linear(self.inner_dim, patch_size * patch_size * self.out_channels, bias=True)
+        self.proj_out = RowParallelLinear(
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/ovis_image/ovis_image_transformer.py]: Using `RowParallelLinear` with `input_is_parallel=False` for the final `proj_out` means every rank does the full matmul and then does an all-reduce — this is more expensive than a plain `nn.Linear`. Since this is the final projection (output is not consumed by any further TP layer), a plain `nn.Linear` would be simpler and equally correct, consistent with the Flux model.

**Comment by hsliuustc0106** [comment]
@vllm-omni-reviewer

**Comment by ZANMANGLOOPYE** [comment]
@lishunyang12  Thanks for the great review and suggestions! I've updated the code to address all your comments. Please take a look when you have a chance.

**Comment by Gaohan123** [comment]
@lishunyang12 PTAL again

**Review by lishunyang12 — COMMENTED** [review]
Thanks for the updates. Verified that most earlier comments are addressed:

- `hidden_states.to(query.dtype)` cast — kept ✓
- `logger.warning` for unexpected params — added ✓  
- Weight layout comment in SwiGLU — added ✓
- `context_embedder`/`x_embedder` — no longer changed to ColumnParallelLinear ✓
- XIELU buffer loading — looks like it was removed because Ovis doesn't use XIELU. Fine.
- `AdaLayerNormZero` custom impl with `ReplicatedLinear` — makes sense for TP ✓

Two things I want to double-check:

1. **`proj_out` in `OvisImageSingleTransformerBlock`** — This is `RowParallelLinear(dim + self.mlp_hidden_dim, dim, input_is_parallel=True)`. The input is `torch.cat([attn_output, mlp_hidden_states], dim=-1)` where `attn_output` has dim `inner_dim / tp_size` and `mlp_hidden_states` has dim `mlp_hidden_dim / tp_size`. So the concat is `(inner_dim + mlp_hidden_dim) / tp_size` which matches `input_is_parallel=True`. This is correct.

2. **`load_weights` manual sharding** — The custom weight loading for `proj_out`, `SwiGLU`, and `proj_mlp` manually splits and re-interleaves weights per TP rank. This is necessary because the concat dimension mixes attention and MLP features. The logic looks correct but is fragile — any change to the model architecture (e.g., different mlp_ratio) would silently produce wrong results. Worth a brief comment in `load_weights` explaining why manual sharding is needed here.

One question: did you verify numerical equivalence between TP=1 and TP=2/4? A pixel-level diff (or LPIPS score) would be more convincing, especially for the SwiGLU weight interleaving.

Overall LGTM after confirming numerical equivalence.

**Comment by Gaohan123** [comment]
Any updates? And please add a UT for it. Thanks!

**Comment by menjiantong** [comment]
Hi @ZANMANGLOOPYE , thanks for starting this work! It looks like it's been inactive for a while. I'd love to help push this across the finish line. Are you still working on this? If not, would you mind if I pick it up from here?

**Comment by ZANMANGLOOPYE** [comment]
Of course, if it is convenient for you to conduct these tests @menjiantong 

> Hi @ZANMANGLOOPYE , thanks for starting this work! It looks like it's been inactive for a while. I'd love to help push this across the finish line. Are you still working on this? If not, would you mind if I pick it up from here?

**Comment by menjiantong** [comment]
> Of course, if it is convenient for you to conduct these tests @menjiantong
> 
> > Hi @ZANMANGLOOPYE , thanks for starting this work! It looks like it's been inactive for a while. I'd love to help push this across the finish line. Are you still working on this? If not, would you mind if I pick it up from here?

Could you please add me as a collaborator to your fork? This will allow me to push commits directly to the branch and continue the work on this PR.

**Comment by wtomin** [comment]
Close this PR as it is duplicate of #4516
